internals/overlord/servstate/manager.go

```diff
@@ -368,6 +372,41 @@ func (m *ServiceManager) WriteMetrics(writer metrics.Writer) error {
 	return nil
 }
 
+// Prune cleans up the in-memory serviceData:
+//   - It removes the serviceData if a service is inactive and its currentSince is older than pruneWait.
+//   - If the number of serviceData entries is still more than maxServiceData, remove inactive services'
+//     serviceData even if they are not older than pruneWait. Inactive services are sorted by currentSince,
+//     and remove the older ones first.
+func (m *ServiceManager) Prune(pruneWait time.Duration, maxServiceData int) {
+	m.servicesLock.Lock()
+	defer m.servicesLock.Unlock()
+
+	now := timeNow()
+	pruneLimit := now.Add(-pruneWait)
+	for name, s := range m.services {
+		if stateToStatus(s.state) == StatusInactive && s.currentSince.Before(pruneLimit) {
+			delete(m.services, name)
+		}
+	}
+
+	if len(m.services) > maxServiceData {
+		var inactive []*serviceData
+		for _, s := range m.services {
+			if stateToStatus(s.state) == StatusInactive {
+				inactive = append(inactive, s)
+			}
+		}
+		slices.SortFunc(inactive, func(a *serviceData, b *serviceData) int {
+			return a.currentSince.Compare(b.currentSince)
+		})
+		excess := len(m.services) - maxServiceData
+		toDelete := min(len(inactive), excess)
+		for i := 0; i < toDelete; i++ {
+			delete(m.services, inactive[i].config.Name)
+		}
```

Comment:
Now that I think about it, "excess" is not really `len(m.services) - maxServiceData` but `len(inactive) - maxServiceData` or specifically `max(len(inactive) - maxServiceData, 0)`. So this seems clearer to me:

```suggestion
		excess := max(len(inactive) - maxServiceData, 0)
		for i := range excess {
			delete(m.services, inactive[i].config.Name)
		}
```

Is that right?

Reply:
Yes, I was thinking maxServiceData is for the total number of serviceData entries, active or not. If maxServiceData means inactive services only, your logic is correct, see my reply in the other comment, and I have made the changes accordingly.